source/YTEPs/YTEP-0036.rst

```diff
@@ -0,0 +1,248 @@
+
+.. code-block:: bash
+
+  $ pytest --with-answer-testing --answer-store --answer-big-data -k "TestArt and test_d9p"
+
+Writing New Tests
+^^^^^^^^^^^^^^^^^
+
+If one is writing tests for a new frontend, one should create a ``tests`` directory inside of the new frontend directory and, within the ``tests`` directory, create a file called ``test_outputs.py``. Within this file, one should define a new class that inherits from ``AnswerTest`` and is marked by pytest as being an answer test. All of the frontend tests should now be methods of this class. For example:
+
+.. code-block:: python
+
+  import pytest
+
+  from yt.utilities.answer_testing import framework as fw
+  from yt.utilities.answer_testing import utils
+
+  # The first decorator tell pytest to skip the tests in this class if the
+  # --with-answer-testing option was not passed on the command line. The
+  # second decorator tells pytest to use the answer_file fixture, which
+  # properly sets TestNewFrontend answer_file
+  @pytest.mark.answer_test
+  @pytest.mark.usefixtures('answer_file')
+  class TestNewFrontend(fw.AnswerTest):
+    # By using the hashing fixture, all of the data stored in test_result
+    # will be hashed upon the method's completion
+    @pytest.mark.usefixtures('hashing')
+    @utils.requires_ds(test_data_file)
+    # The method can be parametrized in the conftest.py file. See below
+    def test_method1(self, param1, param2):
+      ds = utils.data_dir_load(test_data_file)
+      test_result = self.some_answer_test(ds, param1, param2)
+      self.hashes.update({'some_answer_test' : test_result})
+
+If desired, each test method can be individually parametrized by creating a ``conftest.py`` file next to the ``test_outputs.py`` file. The example above would utilize this file like so:
+
+.. code-blocK:: python
+
+  def pytest_generate_tests(metafunc):
+    if metafunc.function.__name__ == 'test_method1':
+      metafunc.parametrize('param1', list_of_param1_values, ids=list_of_param1_ids)
+      metafunc.parametrize('param2', list_of_param2_values, ids=list_of_param2_ids)
+
+This ``pytest_generate_tests`` function will generate a separate test function for each combination of values in the two parameter lists. To parametrize against multiple parameters without having every combination:
+
+.. code-blocK:: python
+
+  def pytest_generate_tests(metafunc):
+    if metafunc.function.__name__ == 'test_method1':
+      metafunc.parametrize('param1, param2', list_of_parameter_combination_tuples, ids=list_of_ids)
+
+That is, one must generate a list whose elements are ``(param1, param2)`` tuples.
+
+If one is writing a new answer test (such as the ``field_values_test``), one should make the test a method of the ``AnswerTest`` class. For example:
+
+.. code-block:: python
+
+  # This is in yt/utilities/answer_testing.framework.py
+  class AnswerTest:
+    # The other methods are here. New method goes below
+    def my_new_answer_test(self, test_param1, test_param2):
```

Comment:
what are the test discovery rules ? I'm wondering if `my_new_answer_test` is actually a "valid" test name according to pytest's default rules.

Reply:
The part of the doc this comment applied too has been removed, but my question hasn't been answered (unless the answer is in the new version, but I'm not looking at it right now. In this case, please resolve this thread @jcoughlin11).